### PR TITLE
feat: support migration of meta database

### PR DESCRIPTION
**Pull request by kan-bayashi**
## What?

- support migration of meta database

## Why?

- to migrate database from the old version

Before

```json
{
  "database_id_df": {
    "1": {
      "database_id": "old",
      "df_name": "db_7bb29237_meta",
      "_uuid": "149603e6c03516362a8da23f624db945",
      "_creation_time": 1669881528.823105
    }
  },
  "--config--database_id_df": {
    "1": {
      "columns": [
        {
          "name": "database_id",
          "dtype": "string",
          "aggregation": "first",
          "display_name": "Database ID"
        },
        {
          "name": "df_name",
          "dtype": "string",
          "aggregation": "first",
          "display_name": "DataFrame name"
        }
      ],
      "index_columns": [
        "database_id"
      ],
      "_uuid": "__config__"
    }
  },
  "db_7bb29237_meta": {
    "1": {
      "description": "Description",
      "record_id": "sample",
      "data_type": "raw_data",
      "path": "test/records/sample/data/records.bag",
      "start_timestamp": 1550125637.22,
      "end_timestamp": 1550125637.53,
      "content_type": "application/rosbag",
      "contents": {
        "/points_concat_downsampled": {
          "msg_type": "sensor_msgs/PointCloud2",
          "msg_md5sum": "1158d486dd51d683ce2f1be655c3c181",
          "count": 4,
          "frequency": 10,
          "tags": [
            "lidar",
            "downsampled"
          ]
        }
      },
      "_kind": "file",
      "_record": "7cc5c0fd8c3601ff09f1c24281cf35ad",
      "_uuid": "d7d651c4efb568bbcafabe73f03ebb57",
      "_creation_time": 1669881528.805312
    },
    "2": {
      "description": "Description",
      "record_id": "csv_model_test",
      "data_type": "raw_data",
      "path": "test/records/csv_model_test/data/test.csv",
      "start_timestamp": 1489728491,
      "end_timestamp": 1489728570.957,
      "content_type": "text/csv",
      "contents": {
        "camera/front-center": {
          "tags": [
            "camera",
            "front",
            "center",
            "timestamps"
          ]
        }
      },
      "_kind": "file",
      "_record": "cdbcdf553baca63623abb1c68a7d0d5a",
      "_uuid": "2b81c94396b3a2df4dc4fb9845a4fbe4",
      "_creation_time": 1669881528.805917
    },
    "3": {
      "description": "json file",
      "database_id": "json datbase",
      "record_id": "test",
      "data_type": "test",
      "path": "test/records/json_model_test/json_test.json",
      "content_type": "application/json",
      "contents": {
        "test": {
          "tags": [
            "test1",
            "test2"
          ]
        }
      },
      "_kind": "file",
      "_record": "c36b825d796e17ebd87b6369d0de4757",
      "_uuid": "5e7efc3a63c02640b42b6a46dd0f6360",
      "_creation_time": 1669881528.806538
    },
    "4": {
      "description": "Forecast",
      "record_id": "test",
      "data_type": "forecast",
      "path": "test/records/forecast_model_test/forecast_test.csv",
      "content_type": "text/csv",
      "contents": {
        "forecast": {
          "tags": [
            "test1",
            "test2"
          ]
        }
      },
      "_kind": "file",
      "_record": "c36b825d796e17ebd87b6369d0de4757",
      "_uuid": "c50d8949f0f07d74070d4e4b29410370",
      "_creation_time": 1669881528.807
    },
    "5": {
      "description": "Description.",
      "record_id": "016_00000000030000000015_1095_01",
      "data_type": "annotation",
      "path": "test/records/annotation_model_test/annotation_test.csv",
      "start_timestamp": 1484628818.02,
      "end_timestamp": 1484628823.98,
      "content_type": "text/csv",
      "contents": {
        "risk_annotation": {
          "tags": [
            "risk_score",
            "scene_description",
            "risk_factor"
          ]
        }
      },
      "_kind": "file",
      "_record": "c0fd5325845ab31db1012545a2cc1126",
      "_uuid": "27ec9a5e953fa5524e2b4feb0ea78edb",
      "_creation_time": 1669881528.807419
    }
  },
  "--config--db_7bb29237_meta": {
    "1": {
      "_hash_digest_size": 4,
      "_df_name": "db_{database_id}_meta",
      "columns": [
        {
          "name": "record_id",
          "dtype": "string",
          "aggregation": "first",
          "display_name": "Record ID"
        },
        {
          "name": "path",
          "dtype": "string",
          "aggregation": "push",
          "display_name": "File path"
        },
        {
          "name": "contents",
          "dtype": "dict",
          "aggregation": "mergeObjects",
          "display_name": "Contents"
        },
        {
          "name": "_record",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "_record"
        },
        {
          "name": "start_timestamp",
          "dtype": "float",
          "aggregation": "first",
          "display_name": "start_timestamp"
        },
        {
          "name": "_kind",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "_kind"
        },
        {
          "name": "description",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "description"
        },
        {
          "name": "content_type",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "content_type"
        },
        {
          "name": "data_type",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "data_type"
        },
        {
          "name": "end_timestamp",
          "dtype": "float",
          "aggregation": "first",
          "display_name": "end_timestamp"
        },
        {
          "name": "database_id",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "database_id"
        }
      ],
      "index_columns": [
        "_kind",
        "record_id",
        "path"
      ],
      "_uuid": "__config__"
    }
  }
}
```

After
```json
{
  "database_id_df": {
    "2": {
      "database_id": "new",
      "df_name": "db_488cd754_meta",
      "_uuid": "22af645d1859cb5ca6da0c484f1f37ea",
      "_creation_time": 1669881573.558427
    }
  },
  "--config--database_id_df": {
    "1": {
      "columns": [
        {
          "name": "database_id",
          "dtype": "string",
          "aggregation": "first",
          "display_name": "Database ID"
        },
        {
          "name": "df_name",
          "dtype": "string",
          "aggregation": "first",
          "display_name": "DataFrame name"
        }
      ],
      "index_columns": [
        "database_id"
      ],
      "_uuid": "__config__"
    }
  },
  "db_488cd754_meta": {
    "1": {
      "d7d651c4efb568bbcafabe73f03ebb57": {
        "description": "Description",
        "record_id": "sample",
        "data_type": "raw_data",
        "path": "test/records/sample/data/records.bag",
        "start_timestamp": 1550125637.22,
        "end_timestamp": 1550125637.53,
        "content_type": "application/rosbag",
        "contents": {
          "/points_concat_downsampled": {
            "msg_type": "sensor_msgs/PointCloud2",
            "msg_md5sum": "1158d486dd51d683ce2f1be655c3c181",
            "count": 4,
            "frequency": 10,
            "tags": [
              "lidar",
              "downsampled"
            ]
          }
        },
        "_kind": "file",
        "_record": "7cc5c0fd8c3601ff09f1c24281cf35ad",
        "_uuid": "d7d651c4efb568bbcafabe73f03ebb57",
        "_creation_time": 1669881528.805312
      },
      "2b81c94396b3a2df4dc4fb9845a4fbe4": {
        "description": "Description",
        "record_id": "csv_model_test",
        "data_type": "raw_data",
        "path": "test/records/csv_model_test/data/test.csv",
        "start_timestamp": 1489728491,
        "end_timestamp": 1489728570.957,
        "content_type": "text/csv",
        "contents": {
          "camera/front-center": {
            "tags": [
              "camera",
              "front",
              "center",
              "timestamps"
            ]
          }
        },
        "_kind": "file",
        "_record": "cdbcdf553baca63623abb1c68a7d0d5a",
        "_uuid": "2b81c94396b3a2df4dc4fb9845a4fbe4",
        "_creation_time": 1669881528.805917
      },
      "5e7efc3a63c02640b42b6a46dd0f6360": {
        "description": "json file",
        "database_id": "json datbase",
        "record_id": "test",
        "data_type": "test",
        "path": "test/records/json_model_test/json_test.json",
        "content_type": "application/json",
        "contents": {
          "test": {
            "tags": [
              "test1",
              "test2"
            ]
          }
        },
        "_kind": "file",
        "_record": "c36b825d796e17ebd87b6369d0de4757",
        "_uuid": "5e7efc3a63c02640b42b6a46dd0f6360",
        "_creation_time": 1669881528.806538
      },
      "c50d8949f0f07d74070d4e4b29410370": {
        "description": "Forecast",
        "record_id": "test",
        "data_type": "forecast",
        "path": "test/records/forecast_model_test/forecast_test.csv",
        "content_type": "text/csv",
        "contents": {
          "forecast": {
            "tags": [
              "test1",
              "test2"
            ]
          }
        },
        "_kind": "file",
        "_record": "c36b825d796e17ebd87b6369d0de4757",
        "_uuid": "c50d8949f0f07d74070d4e4b29410370",
        "_creation_time": 1669881528.807
      },
      "27ec9a5e953fa5524e2b4feb0ea78edb": {
        "description": "Description.",
        "record_id": "016_00000000030000000015_1095_01",
        "data_type": "annotation",
        "path": "test/records/annotation_model_test/annotation_test.csv",
        "start_timestamp": 1484628818.02,
        "end_timestamp": 1484628823.98,
        "content_type": "text/csv",
        "contents": {
          "risk_annotation": {
            "tags": [
              "risk_score",
              "scene_description",
              "risk_factor"
            ]
          }
        },
        "_kind": "file",
        "_record": "c0fd5325845ab31db1012545a2cc1126",
        "_uuid": "27ec9a5e953fa5524e2b4feb0ea78edb",
        "_creation_time": 1669881528.807419
      },
      "_uuid": "d41d8cd98f00b204e9800998ecf8427e",
      "_creation_time": 1669881573.556555
    }
  },
  "--config--db_488cd754_meta": {
    "1": {
      "_hash_digest_size": 4,
      "_df_name": "db_{database_id}_meta",
      "columns": [
        {
          "name": "record_id",
          "dtype": "string",
          "aggregation": "first",
          "display_name": "Record ID"
        },
        {
          "name": "path",
          "dtype": "string",
          "aggregation": "push",
          "display_name": "File path"
        },
        {
          "name": "contents",
          "dtype": "dict",
          "aggregation": "mergeObjects",
          "display_name": "Contents"
        },
        {
          "name": "_record",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "_record"
        },
        {
          "name": "start_timestamp",
          "dtype": "float",
          "aggregation": "first",
          "display_name": "start_timestamp"
        },
        {
          "name": "_kind",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "_kind"
        },
        {
          "name": "description",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "description"
        },
        {
          "name": "content_type",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "content_type"
        },
        {
          "name": "data_type",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "data_type"
        },
        {
          "name": "end_timestamp",
          "dtype": "float",
          "aggregation": "first",
          "display_name": "end_timestamp"
        },
        {
          "name": "database_id",
          "dtype": "str",
          "aggregation": "first",
          "display_name": "database_id"
        },
        {
          "name": "c50d8949f0f07d74070d4e4b29410370",
          "dtype": "Document",
          "aggregation": "first",
          "display_name": "c50d8949f0f07d74070d4e4b29410370"
        },
        {
          "name": "2b81c94396b3a2df4dc4fb9845a4fbe4",
          "dtype": "Document",
          "aggregation": "first",
          "display_name": "2b81c94396b3a2df4dc4fb9845a4fbe4"
        },
        {
          "name": "d7d651c4efb568bbcafabe73f03ebb57",
          "dtype": "Document",
          "aggregation": "first",
          "display_name": "d7d651c4efb568bbcafabe73f03ebb57"
        },
        {
          "name": "5e7efc3a63c02640b42b6a46dd0f6360",
          "dtype": "Document",
          "aggregation": "first",
          "display_name": "5e7efc3a63c02640b42b6a46dd0f6360"
        },
        {
          "name": "27ec9a5e953fa5524e2b4feb0ea78edb",
          "dtype": "Document",
          "aggregation": "first",
          "display_name": "27ec9a5e953fa5524e2b4feb0ea78edb"
        }
      ],
      "index_columns": [
        "_kind",
        "record_id",
        "path"
      ],
      "_uuid": "__config__"
    }
  },
  "db_7bb29237_meta": {},
  "--config--db_7bb29237_meta": {}
}
```

### CODE REVIEW

File: pydtk/db/v4/handlers/meta.py [hunk 1]
```diff
@@ -297,6 +299,43 @@ def save(self, *args, **kwargs):
...
+            new_meta_db_handler._config.__setitem__(k, v, force=True)
+        new_meta_db_handler.save()
+
+        # copy data in old table to new table
+        # TODO(kan-bayashi): Increase limit to perform chunk-wise processing
+        self.read(limit=1)
+        for idx in range(self._count_total):
+            self.read(limit=1, offset=idx)
+            # NOTE(kan-bayashi): Save memory usage
+            new_meta_db_handler = MetaDBHandler(database_id=new_database_id)
```
Comment (d-hayashi) [inline pydtk/db/v4/handlers/meta.py:324]:
```suggestion
            new_meta_db_handler = MetaDBHandler(
              database_id=new_database_id,
              db_engine=self._db_engine,
              db_host=self._db_host,
              db_username=self._db_username,
              db_password=self._db_password,
              db_name=self._db_name,
              base_dir_path=self.base_dir_path,
          )
```

File: test/test_db.py [hunk 2]
```diff
@@ -101,6 +101,53 @@ def test_create_db(
...
+        base_dir_path=os.path.join(os.getcwd(), "test"),
+    )
+    handler.read()
+    assert isinstance(handler, V4MetaDBHandler)
+    _add_files_to_db(handler)
+    handler.rename_database_id("new")
+
+    # check old one is empty
+    old_handler = V4MetaDBHandler(
+        database_id="old",
```
Comment (d-hayashi) [inline test/test_db.py:139]:
```suggestion
        database_id="old",
        db_class='meta',
        db_engine=db_engine,
        db_host=db_host,
        db_username=db_username,
        db_password=db_password,
        db_name=db_name,
        base_dir_path=os.path.join(os.getcwd(), "test"),
```

File: test/test_db.py [hunk 3]
```diff
@@ -101,6 +101,53 @@ def test_create_db(
...
+    )
+    old_handler.read()
+    assert len(old_handler.data) == 0
+
+    new_handler = V4MetaDBHandler(
+        database_id="new",
```
Comment (d-hayashi) [inline test/test_db.py:145]:
```suggestion
        database_id="new",
        db_class='meta',
        db_engine=db_engine,
        db_host=db_host,
        db_username=db_username,
        db_password=db_password,
        db_name=db_name,
        base_dir_path=os.path.join(os.getcwd(), "test"),
```

Comment (kan-bayashi) [reply test/test_db.py:151]: fixed

File: test/test_db.py (hunk 2 above)
Comment (kan-bayashi) [reply test/test_db.py:139]: fixed

File: pydtk/db/v4/handlers/meta.py (hunk 1 above)
Comment (kan-bayashi) [reply pydtk/db/v4/handlers/meta.py]: fixed

**Review by d-hayashi — APPROVED** [review]
LGTM!

**Pull request merged**